pydeseq2/dds.py

```diff
@@ -793,3 +816,84 @@ def _refit_without_outliers(
             new_all_zeroes.sum()
         )
         self[:, self.new_all_zeroes_genes].varm["LFC"] = np.zeros(new_all_zeroes.sum())
+
+    def _fit_iterate_size_factors(self, niter: int = 10, quant: float = 0.95) -> None:
+        """
+        Fit size factors using the ``iterative`` method.
+
+        Used when each gene has at least one zero.
+
+        Parameters
+        ----------
+        niter : int
+            Maximum number of iterations to perform (default: ``10``).
+
+        quant : float
+            Quantile value at which negative likelihood is cut in the optimization
+            (default: ``0.95``).
+
+        """
+
+        self.obsm["size_factors"] = np.ones(self.n_obs)
+
+        # Reduce the design matrix to an intercept and reconstruct at the end
+        self.obsm["design_matrix_buffer"] = self.obsm["design_matrix"].copy()
+        self.obsm["design_matrix"] = pd.DataFrame(
+            1, index=self.obs_names, columns=[["intercept"]]
+        )
+
+        for i in range(niter):
+            # Estimate dispersions based on current size factors
+            self.fit_genewise_dispersions()
+
+            # Use a mean trend curve
+            use_for_mean_genes = self.var_names[
+                (self.varm["genewise_dispersions"] > 10 * self.min_disp)
+                & self.varm["non_zero"]
+            ]
+
+            mean_disp = trimmed_mean(
+                self[:, use_for_mean_genes].varm["genewise_dispersions"], trim=0.001
+            )
+            self.varm["fitted_dispersions"] = np.ones(self.n_vars) * mean_disp
+            self.fit_dispersion_prior()
+            self.fit_MAP_dispersions()
+            old_sf = self.obsm["size_factors"].copy()
+
+            # Fit size factors using MLE
+            def fn(p):
```

Comment:
I am no sure if I follow why this function is defined within the for loop? Could it be defined outside and perhaps made private by adding '_' or is there are specific reason for having it here? and if there is a reason could you add a short comment on a why?

Reply:
I've moved this function outside the loop